File: tzio-core/scala/src/main/scala/com/kineolyan/tzio/v1/scala/refs/OutputAdapter.scala
package com.kineolyan.tzio.v1.scala.refs

import com.kineolyan.tzio.v1.api.ref

/**
  * Object adapting definition of outputs to their actual types in the Scala core
  */
object OutputAdapter extends ref.OutputReferenceVisitor[OutputReference] {

  def convert(definition: ref.OutputReferenceType): OutputReference =
    definition.accept(this)

  override def visit(reference: ref.SlotReference): OutputReference =
    OutSlotReference(reference.slot)

  override def visit(reference: ref.AccReference): OutputReference =
    OutAccReference()

  override def visit(reference: ref.NilReference): OutputReference =
    OutNilReference()
}

File: tzio-core/jdk18/src/main/java/com/kineolyan/tzio/v1/java/ref/OutputAdapter.java
package com.kineolyan.tzio.v1.java.ref;

import com.kineolyan.tzio.v1.api.ref.AccReference;
import com.kineolyan.tzio.v1.api.ref.NilReference;
import com.kineolyan.tzio.v1.api.ref.OutputReferenceType;
import com.kineolyan.tzio.v1.api.ref.OutputReferenceVisitor;
import com.kineolyan.tzio.v1.api.ref.SlotReference;

public class OutputAdapter implements OutputReferenceVisitor<OutputReference> {

	public OutputReference convert(final OutputReferenceType type) {
		return switch (type) {
			case SlotReference slot -> visit(slot);
			case NilReference nil -> visit(nil);
			case AccReference acc -> visit(acc);
		};
	}

	@Override
	public OutputReference visit(final SlotReference ref) {
		return References.outSlot(ref.slot());
	}

	@Override
	public OutputReference visit(final AccReference ref) {
		return References.acc();
	}

	@Override
	public OutputReference visit(final NilReference ref) {
		return References.outNil();
	}

}
